File: backend/payment-service/app/services/payment_service.py
import logging
import random
import string
import time
from typing import Dict, Any

from sqlalchemy.ext.asyncio import AsyncSession

from app.events.kafka_producer import PaymentEventPublisher
from app.models.payment import Payment, PaymentStatus
from app.repositories.payment_repository import PaymentRepository
from app.repositories.payment_audit_log_repository import PaymentAuditLogRepository
from app.schemas.payment import PaymentRequest, PaymentResponse
from app.services.gateway_adapter import MockPaymentGatewayAdapter

logger = logging.getLogger(__name__)


class PaymentAlreadyProcessedException(Exception):
    pass


class PaymentNotFoundException(Exception):
    pass


class PaymentService:
    def __init__(
        self,
        db: AsyncSession,
        event_publisher: PaymentEventPublisher,
        gateway_adapter: MockPaymentGatewayAdapter,
    ):
        self.payment_repo = PaymentRepository(db)
        self.audit_repo = PaymentAuditLogRepository(db)
        self.event_publisher = event_publisher
        self.gateway_adapter = gateway_adapter

    async def initiate_payment(self, request: PaymentRequest) -> PaymentResponse:
        logger.info(f"Initiating payment for booking ID: {request.booking_id}")

        # Check if payment already exists for this booking
        if await self.payment_repo.exists_by_booking_id(request.booking_id):
            raise PaymentAlreadyProcessedException("Payment already exists for this booking")

        # Create payment record
        payment = Payment(
            payment_reference=self._generate_payment_reference(),
            booking_id=request.booking_id,
            user_id=request.user_id,
            amount=request.amount,
            currency="INR",
            payment_method=request.payment_method,
            payment_status=PaymentStatus.INITIATED,
            gateway_name=request.gateway_name,
        )

        saved_payment = await self.payment_repo.save(payment)

        # Initiate payment with gateway
        gateway_response = await self.gateway_adapter.initiate_payment(
            saved_payment.payment_reference, saved_payment.amount, request.payment_method
        )

        # Update payment with gateway transaction ID
        saved_payment.gateway_transaction_id = gateway_response.transaction_id
        await self.payment_repo.update(saved_payment)

        # Log audit event
        await self.audit_repo.create(
            saved_payment.id,
            "PAYMENT_INITIATED",
            {
                "paymentReference": saved_payment.payment_reference,
                "amount": str(saved_payment.amount),
                "gatewayTransactionId": gateway_response.transaction_id or "N/A",
            },
        )

        # Publish Kafka event
        await self.event_publisher.publish_payment_initiated(saved_payment)

        logger.info(f"Payment initiated successfully: {saved_payment.payment_reference}")

        # Create response
        response = PaymentResponse.model_validate(saved_payment)
        response.payment_url = gateway_response.payment_url
        return response

    async def confirm_payment(self, payment_reference: str, gateway_transaction_id: str):
        logger.info(f"Confirming payment: {payment_reference}")

        payment = await self.payment_repo.find_by_payment_reference(payment_reference)
        if not payment:
            raise PaymentNotFoundException(f"Payment not found: {payment_reference}")

        # Idempotent - if already confirmed, just return
        if payment.payment_status == PaymentStatus.SUCCESS:
            logger.warning(f"Payment already confirmed: {payment_reference}")
            return

        payment.payment_status = PaymentStatus.SUCCESS
        payment.gateway_transaction_id = gateway_transaction_id
        await self.payment_repo.update(payment)

        # Log audit event
        await self.audit_repo.create(
            payment.id,
            "PAYMENT_CONFIRMED",
            {
                "paymentReference": payment_reference,
                "gatewayTransactionId": gateway_transaction_id,
            },
        )

        # Publish Kafka event
        await self.event_publisher.publish_payment_success(payment)

        logger.info(f"Payment confirmed successfully: {payment_reference}")

    async def fail_payment(self, payment_reference: str, failure_reason: str):
        logger.info(f"Failing payment: {payment_reference}")

        payment = await self.payment_repo.find_by_payment_reference(payment_reference)
        if not payment:
            raise PaymentNotFoundException(f"Payment not found: {payment_reference}")

        payment.payment_status = PaymentStatus.FAILED
        payment.failure_reason = failure_reason
        await self.payment_repo.update(payment)

        # Log audit event
        await self.audit_repo.create(
            payment.id,
            "PAYMENT_FAILED",
            {"paymentReference": payment_reference, "failureReason": failure_reason},
        )

        # Publish Kafka event
        await self.event_publisher.publish_payment_failed(payment)

        logger.error(f"Payment failed: {payment_reference}, reason: {failure_reason}")

    async def get_payment_by_id(self, payment_id: int) -> PaymentResponse:
        payment = await self.payment_repo.find_by_id(payment_id)
        if not payment:
            raise PaymentNotFoundException(f"Payment not found with id: {payment_id}")
        return PaymentResponse.model_validate(payment)

    async def get_payment_by_booking_id(self, booking_id: int) -> PaymentResponse:
        payment = await self.payment_repo.find_by_booking_id(booking_id)
        if not payment:
            raise PaymentNotFoundException(f"Payment not found for booking: {booking_id}")
        return PaymentResponse.model_validate(payment)

    def _generate_payment_reference(self) -> str:
        timestamp = int(time.time() * 1000)
        random_str = "".join(random.choices(string.ascii_uppercase + string.digits, k=8))
        return f"PAY{timestamp}{random_str}"

File: backend/payment-service/src/main/java/com/bookmyshow/payment/service/PaymentService.java
package com.bookmyshow.payment.service;

import com.bookmyshow.payment.dto.GatewayResponse;
import com.bookmyshow.payment.dto.PaymentRequest;
import com.bookmyshow.payment.dto.PaymentResponse;
import com.bookmyshow.payment.exception.PaymentAlreadyProcessedException;
import com.bookmyshow.payment.exception.PaymentNotFoundException;
import com.bookmyshow.payment.gateway.MockPaymentGatewayAdapter;
import com.bookmyshow.payment.kafka.PaymentEventPublisher;
import com.bookmyshow.payment.model.Payment;
import com.bookmyshow.payment.model.PaymentAuditLog;
import com.bookmyshow.payment.repository.PaymentAuditLogRepository;
import com.bookmyshow.payment.repository.PaymentRepository;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

@Service
@Slf4j
@Transactional
public class PaymentService {

    @Autowired
    private PaymentRepository paymentRepository;

    @Autowired
    private PaymentAuditLogRepository auditLogRepository;

    @Autowired
    private MockPaymentGatewayAdapter gatewayAdapter;

    @Autowired
    private PaymentEventPublisher eventPublisher;

    public PaymentResponse initiatePayment(PaymentRequest request) {
        log.info("Initiating payment for booking ID: {}", request.getBookingId());

        if (paymentRepository.existsByBookingId(request.getBookingId())) {
            throw new PaymentAlreadyProcessedException("Payment already exists for this booking");
        }

        Payment payment = new Payment();
        payment.setPaymentReference(generatePaymentReference());
        payment.setBookingId(request.getBookingId());
        payment.setUserId(request.getUserId());
        payment.setAmount(request.getAmount());
        payment.setCurrency("INR");
        payment.setPaymentMethod(request.getPaymentMethod());
        payment.setPaymentStatus(Payment.PaymentStatus.INITIATED);
        payment.setGatewayName(request.getGatewayName());

        Payment savedPayment = paymentRepository.save(payment);

        GatewayResponse gatewayResponse = gatewayAdapter.initiatePayment(
                savedPayment.getPaymentReference(),
                savedPayment.getAmount(),
                request.getPaymentMethod()
        );

        savedPayment.setGatewayTransactionId(gatewayResponse.getTransactionId());
        paymentRepository.save(savedPayment);

        logAuditEvent(savedPayment.getId(), "PAYMENT_INITIATED", Map.of(
                "paymentReference", savedPayment.getPaymentReference(),
                "amount", savedPayment.getAmount().toString(),
                "gatewayTransactionId", gatewayResponse.getTransactionId() != null ? gatewayResponse.getTransactionId() : "N/A"
        ));

        eventPublisher.publishPaymentInitiated(savedPayment);

        log.info("Payment initiated successfully: {}", savedPayment.getPaymentReference());

        PaymentResponse response = PaymentResponse.from(savedPayment);
        response.setPaymentUrl(gatewayResponse.getPaymentUrl());
        return response;
    }

    public void confirmPayment(String paymentReference, String gatewayTransactionId) {
        log.info("Confirming payment: {}", paymentReference);

        Payment payment = paymentRepository.findByPaymentReference(paymentReference)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found: " + paymentReference));

        if (payment.getPaymentStatus() == Payment.PaymentStatus.SUCCESS) {
            log.warn("Payment already confirmed: {}", paymentReference);
            return;
        }

        payment.setPaymentStatus(Payment.PaymentStatus.SUCCESS);
        payment.setGatewayTransactionId(gatewayTransactionId);
        paymentRepository.save(payment);

        logAuditEvent(payment.getId(), "PAYMENT_CONFIRMED", Map.of(
                "paymentReference", paymentReference,
                "gatewayTransactionId", gatewayTransactionId
        ));

        eventPublisher.publishPaymentSuccess(payment);

        log.info("Payment confirmed successfully: {}", paymentReference);
    }

    public void failPayment(String paymentReference, String failureReason) {
        log.info("Failing payment: {}", paymentReference);

        Payment payment = paymentRepository.findByPaymentReference(paymentReference)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found: " + paymentReference));

        payment.setPaymentStatus(Payment.PaymentStatus.FAILED);
        payment.setFailureReason(failureReason);
        paymentRepository.save(payment);

        logAuditEvent(payment.getId(), "PAYMENT_FAILED", Map.of(
                "paymentReference", paymentReference,
                "failureReason", failureReason
        ));

        eventPublisher.publishPaymentFailed(payment);

        log.error("Payment failed: {}, reason: {}", paymentReference, failureReason);
    }

    public PaymentResponse getPaymentById(Long id) {
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found with id: " + id));
        return PaymentResponse.from(payment);
    }

    public PaymentResponse getPaymentByBookingId(Long bookingId) {
        Payment payment = paymentRepository.findByBookingId(bookingId)
                .orElseThrow(() -> new PaymentNotFoundException("Payment not found for booking: " + bookingId));
        return PaymentResponse.from(payment);
    }

    private String generatePaymentReference() {
        return "PAY" + System.currentTimeMillis() + RandomStringUtils.randomAlphanumeric(8).toUpperCase();
    }

    private void logAuditEvent(Long paymentId, String eventType, Map<String, Object> eventData) {
        PaymentAuditLog auditLog = new PaymentAuditLog();
        auditLog.setPaymentId(paymentId);
        auditLog.setEventType(eventType);
        auditLog.setEventData(new HashMap<>(eventData));
        auditLogRepository.save(auditLog);
    }
}
